lens/lotus/lotus.go

```diff
@@ -6,60 +6,74 @@ import (
 	"net/http"
 	"strings"
 
-	"github.com/filecoin-project/sentinel-visor/lens"
-
+	"github.com/filecoin-project/go-jsonrpc"
+	lotus_api "github.com/filecoin-project/lotus/api"
+	"github.com/filecoin-project/lotus/api/client"
+	lcli "github.com/filecoin-project/lotus/cli"
+	lru "github.com/hashicorp/golang-lru"
 	logging "github.com/ipfs/go-log/v2"
 	ma "github.com/multiformats/go-multiaddr"
 	manet "github.com/multiformats/go-multiaddr/net"
 	"github.com/urfave/cli/v2"
 	"golang.org/x/xerrors"
 
-	"github.com/filecoin-project/go-jsonrpc"
-	lotus_api "github.com/filecoin-project/lotus/api"
-	"github.com/filecoin-project/lotus/api/client"
-	lcli "github.com/filecoin-project/lotus/cli"
+	"github.com/filecoin-project/sentinel-visor/lens"
 )
 
 var log = logging.Logger("visor/lens/lotus")
 
-func GetFullNodeAPI(cctx *cli.Context) (context.Context, lens.API, lens.APICloser, error) {
+type APIOpener struct {
+	// TODO: replace dependency on cli.Context with tokenMaddr and repo path
+	cctx  *cli.Context
```

Comment:
if open takes a context, why do we need a context on NewAPIOpener? we can take it for the length of opening, but i'm not seeing what the value of storing it on the opener struct is

Reply:
Are you referring to the cli Context?  It's stored because we pass it to visor which adds a tracing key and (now I look closely) sets up a signal handler. I don't know why we're doing this. I'd prefer not to return the context at all from NewAPIOpener but I was trying to limit the scope of this change to connection handling.

Reply:
more that you also pass a ctx to the `Open` function, and that's the one that should probably get tagged / live for the length of the open connection rather than the one passed in constructing the opener object itself.

Reply:
I'm note sure I follow

NewAPIOpener accepts a cli.Context (and uses it for reading flags/settings),  Open accepts a context.Context (and uses it for calls to the lotus API)

Reply:
i think it's just the disjoint between the singleton APIOpener lifetime of the structs vs the duration of connections from the `Open` call where the semantics of how the cctx on the opener object are used is non-obvious.

Reply:
cctx is required (unfortunately) because we have a hard dependency on passing a `*cli.Context` to `github.com/filecoin-project/lotus/cli.GetFullNodeAPI` where it expects to be able to read flags and environment to locate the repo on disk.  We could unpick that dependency.

Reply:
Updated so we don't return the context from the creation of api openers. Does this address your concerns @willscott